jicofo/src/main/kotlin/org/jitsi/jicofo/util/PreferenceAggregator.kt

```diff
@@ -0,0 +1,114 @@
+/*
+ * Jicofo, the Jitsi Conference Focus.
+ *
+ * Copyright @ 2015-Present 8x8, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.jicofo.util
+
+/** Aggregate lists of preferences coming from a large group of people, such that the resulting aggregated
+ * list consists of preference items supported by everyone, and in a rough consensus of preference order.
+ *
+ * The intended use case is maintaining the list of supported codecs for conference visitors.
+ *
+ * Preferences are aggregated using the Borda count; this isn't theoretically optimal, but it's computationally cheap.
+ */
+class PreferenceAggregator(
+    private val onChanged: (List<String>) -> Unit
+) {
+    private val lock = Any()
+
+    var aggregate: List<String> = emptyList()
+        private set
+
+    var count = 0
+        private set
+
+    private val values = mutableMapOf<String, ValueInfo>()
+
+    /**
+     * Add a preference to the aggregator.
+     */
+    fun addPreference(prefs: List<String>) {
+        synchronized(lock) {
+            count++
+            prefs.forEachIndexed { index, element ->
+                val info = values.computeIfAbsent(element) { ValueInfo() }
+                info.count++
+                info.rankAggregate += index
+            }
+            updateAggregate()
+        }
+    }
+
+    /**
+     * Remove a preference from the aggregator.
+     */
+    fun removePreference(prefs: List<String>) {
+        synchronized(lock) {
+            count--
+            check(count >= 0) {
```

Comment:
This could also be force to fail by sending initial presence with no codecs, then updating with some codecs, then leaving repeatedly